### ISSUE TITLE
Compile times scale with number of templates

### ISSUE THREAD

**Issue by wrapperup** [djc/askama#826]
Hi, I'm enjoying askama, it's fantastic. I noticed that compile times for templates grow considerably if you have many templates, which I don't think is too surprising considering it's a proc-macro.

I thought of having templates cached in the user's project (under a `.askama` directory) without using a build script. But then I saw https://github.com/djc/askama/pull/689, but it doesn't seem to have gone anywhere. Has there been any plans or additional thoughts on that to continue on this work?

IMO, it would be great if it could be done purely in the proc-macro. In my limited testing, just skipping the compile step already sped things up considerably, so it seems like the overhead of invoking and parsing a proc-macro wasn't big enough to notice (testing with about 100 templates). I saw there was some unstable Rust features for tracking files in proc-macros, which would be useful for this, but who knows how long that would take to stabilize.

**Comment by djc** [comment]
> The main issue is when making incremental changes, askama will recompile all templates, even if they didn't change. This hurts iteration times a bit.

I find this surprising. Have you verified this? Do you understand why this is the case?

The work in #689 hasn't gone anywhere as far as I know. I still think the code generation at test time would be the optimal approach, but if you have concrete benchmark numbers to demonstrate the benefits of a caching approach I'd be open to considering that as well.

**Comment by wrapperup** [comment]
So, my test implements 50 templates templates in a single module.
```rust
#[derive(Template)]
#[template(path = "index_001.html")]
struct Index001 {
   test: String,
}

...

#[derive(Template)]
#[template(path = "index_050.html")]
struct Index050 {
   test: String,
}
```
Making a change to one or more of these templates causes the compile time to be about ~3.10s.
I ran self profile to get a flamegraph, and curiously the proc-macro stage was the longest.

![image](https://github.com/djc/askama/assets/7478134/ab68c434-9c2f-406d-aa81-1c914a04c1d8)

I also stubbed out `include_bytes!` to see if maybe there was some side-effects with file tracking that may have caused all the macros to run again, but that didn't seem to help much. (I got similar results to above)

I made a very quick-and-dirty prototype (hashes entire proc-macro AST, but not template itself), and the build time dropped to 0.43s for one template, and about 1.78s for compiling all the templates.

Flamegraph for one template change with the cache (much better!):
![image](https://github.com/djc/askama/assets/7478134/f2632028-9be2-49bd-b6c2-d47f479662fc)

and all:
![image](https://github.com/djc/askama/assets/7478134/07d652cd-661d-411b-b6ee-62c165c76ff5)

Honestly, I have no idea why its faster when compiling all templates with cache enabled, that part doesn't really make sense to me. If you're curious, I put the implementation prototype here. It's definitely a draft though: https://github.com/wrapperup/askama/tree/cache-templates

**Comment by djc** [comment]
Okay, so in a crate with 100 templates what is the time for running `cargo check` after touching 1 template in the with cache vs the without cache case?

**Comment by wrapperup** [comment]
I get similar results. Without cache, changing 1 template takes ~2.80s. With cache, it takes ~0.20 for 1, and ~2.82s for all

Here's my test repo: https://github.com/wrapperup/askama-macro-benchmark

**Comment by djc** [comment]
Okay, let's have a PR for a cache. Can you somehow include the template contents in the freshness calculation?

**Comment by wrapperup** [comment]
Yep! This needs a bit of cleanup, but I can probably have something soonish.
